cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/IntraBrokerDiskUsageDistributionGoal.java

```diff
@@ -0,0 +1,520 @@
+        if (rebalanceByMovingLoadOut(disk, clusterModel, optimizedGoals, optimizationOptions)) {
+          rebalanceBySwappingLoadOut(disk, clusterModel, optimizedGoals, optimizationOptions);
+        }
+      }
+      if (diskUtilizationPercentage(disk) < lowerLimit) {
+        if (rebalanceByMovingLoadIn(disk, clusterModel, optimizedGoals, optimizationOptions)) {
+          rebalanceBySwappingLoadIn(disk, clusterModel, optimizedGoals, optimizationOptions);
+        }
+      }
+    }
+  }
+
+  private boolean rebalanceByMovingLoadIn(Disk disk,
+                                          ClusterModel clusterModel,
+                                          Set<Goal> optimizedGoals,
+                                          OptimizationOptions optimizationOptions) {
+    Broker broker = disk.broker();
+    double brokerUtilization = averageDiskUtilizationPercentage(broker);
+    Set<String> excludedTopics = optimizationOptions.excludedTopics();
+
+    PriorityQueue<Disk> candidateDiskPQ = new PriorityQueue<>(
+        (d1, d2) -> Double.compare(diskUtilizationPercentage(d2), diskUtilizationPercentage(d1)));
+    for (Disk candidateDisk : broker.disk()) {
+      // Get candidate disk on broker to try moving load from -- sorted in the order of trial (descending load).
+      if (candidateDisk.state() == Disk.State.ALIVE && diskUtilizationPercentage(candidateDisk) > brokerUtilization) {
+        candidateDiskPQ.add(candidateDisk);
+      }
+    }
+
+    while (!candidateDiskPQ.isEmpty()) {
+      Disk candidateDisk = candidateDiskPQ.poll();
+      candidateDisk.trackSortedReplicas(name(),
+                                        ReplicaSortFunctionFactory.getSelectOnlineReplicas(),
+                                        ReplicaSortFunctionFactory.deprioritizeDiskImmigrants(),
+                                        ReplicaSortFunctionFactory.sortByMetricGroupValue(RESOURCE.name()));
+      for (Iterator<Replica> iterator = candidateDisk.trackedSortedReplicas(name()).reverselySortedReplicas().iterator();
+          iterator.hasNext(); ) {
+        Replica replica = iterator.next();
+        if (excludedTopics.contains(replica.topicPartition().topic())) {
+          continue;
+        }
+        Disk d = maybeMoveReplicaBetweenDisk(clusterModel, replica, Collections.singleton(disk), optimizedGoals);
+        // Only need to check status if the action is taken. This will also handle the case that the source disk
+        // has nothing to move in. In that case we will never re-enqueue that source disk.
+        if (d != null) {
+          if (diskUtilizationPercentage(disk) > _balanceLowerThresholdByBroker.get(broker)) {
+            candidateDisk.untrackSortedReplicas(name());
+            return false;
+          }
+          iterator.remove();
+          // If the source disk has a lower utilization than the next disk in the candidate disk priority queue,
+          // we re-enqueue the source disk and switch to the next disk.
```

Comment:
do we want a rebalanceByMovingLoadOut timeout similar to what's done in `remainingPerDiskSwapTimeMs()` inside `rebalanceBySwappingLoadOut()`?

Reply:
The main reason we want to have a timeout in Swap method it because the time complexity of that operation is too high. For Move method, the time complexity is O(num_of_broker * num_of_replica_on_broker) and this is acceptable based on our experience running Cruise Control. For Swap method, the time complexity is O(num_of_broker * num_of_replica_on_broker * num_of_replica_on_broker).